Silent installer fix

kilobyte1024 opened the pull request:
<img width="556" height="357" alt="image" src="https://github.com/user-attachments/assets/844570c1-0fda-41f9-9303-be3746f26c06" />
Unfortunately I'm getting a Javascript error when starting Discord with this candidate fix. So it will need some more work before being merged in.

kilobyte1024 committed:
draft fix for installer error

Charles-Panchyshyn committed:
fix: correct negation operator for Test-Path

Replaced the incorrect use of `!Test-Path` with `-not (Test-Path ...)` for proper PowerShell syntax. This resolves the error where `!` was not recognized as a valid negation operator in the script.

Charles-Panchyshyn committed:
chore: update .gitignore for VSCode

- Added .vscode folder to .gitignore to prevent local VSCode configuration files from being tracked.

Charles-Panchyshyn committed:
feat: detect version in installer.db to indicate InconsistentInstallerState

- Enhanced logic to parse `installer.db` for version information using a regex pattern.
- Ensured that `installer.db` is only deleted if it does NOT indicate `InconsistentInstallerState`.
- Improved logging to provide detailed feedback on detected version and actions taken.

Charles-Panchyshyn committed:
remove: copy installer.db logic

Removed logic for detecting and handling the latest app version folder in preparation for a replacement implementation. This change simplifies the existing script to allow for a more robust solution.

Charles-Panchyshyn committed:
feat: run squirrel-firstrun silently via S4U task

Added logic to handle and repair the `squirrel-firstrun` state for Discord.

- Checks if the `squirrelFirstRunCompleted` variable is false and initiates a repair process if necessary.
- Generates a temporary scheduled task using a dynamically created XML file.
- Runs the `Discord.exe --squirrel-firstrun` command silently, logging output to a custom log file.
- Monitors the log file for a success message with a 60-second timeout.
- Cleans up temporary files and deletes the scheduled task after execution.
- Outputs appropriate success or warning messages based on the outcome.

This change ensures that Discord's installer state is repaired and can be launched normally.

Charles-Panchyshyn committed:
chore: comments and output string

- Added some comments
- Changed a string to indicate the version source

Charles-Panchyshyn committed:
moved: app version folder detection

- Moved the app version detection logic to the beginning of the script.

Charles-Panchyshyn committed:
chore: Moved custom log clean-up

Charles-Panchyshyn commented:
Thanks for testing this and for the detailed feedback—and for catching that the repair only runs on upgrade; that’s a real gap for fresh installs.

To narrow down the JavaScript error and the S4U/task behavior on your side, it would help to understand how you’re testing:

1. **Testing process**  
   Could you outline how you usually test Chocolatey packages (e.g. fresh install vs upgrade, which steps you run, whether you use `choco install` / `choco upgrade`, and from what kind of prompt/session)?

2. **Environment / VM**  
   The stack trace you shared showed paths under `C:\Users\vagrant\`, which suggests you might be using a Vagrant (or similar) VM for testing. Are you testing in a specific VM or image (e.g. Vagrant box, GitHub Actions runner, other)? S4U scheduled tasks and first-run behavior can differ between physical machines, VMs, and headless/automated environments, so knowing your setup would help a lot.

If you can share that—and any other details about your test flow (e.g. how you install/upgrade, whether you ever run Discord interactively in that environment)—I can try to align the fix with your workflow and see if we can reproduce or work around the JSON error in that context.

kilobyte1024 commented:
> Thanks for testing this and for the detailed feedback—and for catching that the repair only runs on upgrade; that’s a real gap for fresh installs.
> 
> To narrow down the JavaScript error and the S4U/task behavior on your side, it would help to understand how you’re testing:
> 
> 1. **Testing process**  
>    Could you outline how you usually test Chocolatey packages (e.g. fresh install vs upgrade, which steps you run, whether you use `choco install` / `choco upgrade`, and from what kind of prompt/session)?
> 
> 2. **Environment / VM**  
>    The stack trace you shared showed paths under `C:\Users\vagrant\`, which suggests you might be using a Vagrant (or similar) VM for testing. Are you testing in a specific VM or image (e.g. Vagrant box, GitHub Actions runner, other)? S4U scheduled tasks and first-run behavior can differ between physical machines, VMs, and headless/automated environments, so knowing your setup would help a lot.
> 
> If you can share that—and any other details about your test flow (e.g. how you install/upgrade, whether you ever run Discord interactively in that environment)—I can try to align the fix with your workflow and see if we can reproduce or work around the JSON error in that context.

The test steps are the standard steps for chocolatey using chocolatey-testenv. They are well documented in the choco docs and far too in depth to cover in a comment.

I would like to see the test results log or at least an attestation that you tried a fresh install before taking my time to retest another pr please.

But I appreciate you taking the time to try to work around this bug in the Spotify silent installer.

Charles-Panchyshyn committed:
fix: Update success message string in fixOnUpgrade.ps1 to indicate completion of the updater window

Discord's updater no longer logs "CDM completed with status: cdm-ready-".
Use "window.closed win1" instead, which indicates the updater window has closed and we're done checking for updates.

Charles-Panchyshyn committed:
refactor: Replace fixOnUpgrade.ps1 with fixInstallerState.ps1 to ensure consistent installer state on every installation

Charles-Panchyshyn committed:
fix: Update path resolution for fixInstallerState.ps1 to improve script reliability

Charles-Panchyshyn commented:
Refer to [my new PR](https://github.com/kilobyte1024/chocolatey-packages/pull/8)
